### PR TITLE
unit tests with mock i2c

### PR DESCRIPTION
**Pull request by m-mcgowan**
Adds tests over a mock i2c connection. 

The tests are factored into common tests for any transport, plus transport-specific tests for mocked i2c and mocked serial.

DRY's up the python commands in github actions by reusing the Makefile.

### CODE REVIEW

File: test/test_notecard.py [hunk 1]
```diff
@@ -1,13 +1,13 @@
 import os
 import sys
-import serial
-import periphery
+import unittest
+
 import pytest
```
Comment (bsatrom) [inline test/test_notecard.py:5]: Why are you importing `unittest` when `pytest` is already present?

**Comment by bsatrom** [comment]
@m-mcgowan I'm having a hard time seeing what has actually changed in this pr so I am inclined to reject it until it's either simplified or broken into multiple PRs. Rather than taking on a giant refactor of how the tests work, why not add the I2C mocks and tests and we can worry about a refactor later, if at all?

**Comment by m-mcgowan** [comment]
I didn't think it was that big a refactor. All I did was factor out the notecard connection from the tests by moving them in to a base class which calls `get_port()` to retrieve the port with an expected response, and this is used by two subclasses, one for i2c and one for serial that defines `get_port()`.  

It's not presently building on github actions because that's a linux environment and the notecard communications code has the I2C setup hard-coded for to be different for that. It worked on my machine because I'm on OSX - I need to change the logic to also use the same in a test environment so that it is consistent regardless of runtime platform.

When we have HIL testing for this, all we have to do is define another class that uses a real I2C/serial connection running the same tests.

File: test/test_notecard.py (hunk 1 above)
Comment (m-mcgowan) [reply test/test_notecard.py:5]: `unittest` supports classes of tests, which I use for reuse of common tests and parameterization of these across different environments.

File: .github/workflows/manual-run.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ jobs:
     - name: Lint with flake8
       run: |
         # stop the build if there are Python syntax errors or undefined names
-        flake8 test/ notecard/ examples/ --count --ignore=E722,F401,F403,W503 --show-source --statistics
+        flake8 test/ notecard/ examples/ --count --ignore=E722,F401,F403,W503,E501 --show-source --statistics
```
Comment (haydenroche5) [inline .github/workflows/manual-run.yml:37]: Can we add a comment explaining what we're ignoring?

**Comment by m-mcgowan** [comment]
[mock tests for note-python and I2C](https://trello.com/c/lGskUR4t/11-mock-tests-for-note-python-and-i2c)

File: test/test_notecard.py [hunk 3]
```diff
@@ -34,449 +33,385 @@ def get_i2c_and_port():
...
+        nCard, _ = self.get_port()
+        userAgent = nCard.GetUserAgent()
 
-def test_file_pendingChanges():
-    nCard, port = get_serial_and_port()
+        assert userAgent['req_interface'] == 'i2c'
+        assert userAgent['req_port'] is not None
 
-    port.readline.return_value = "{\"changes\":1}\r\n"
+    def test_debug_mode_on_i2c(self):
```
Comment (haydenroche5) [inline test/test_notecard.py:410]: The first several lines of this look like what `get_i2c_and_port` does. Can we use that instead?

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: .github/workflows/manual-run.yml (hunk 2 above)
Comment (m-mcgowan) [reply .github/workflows/manual-run.yml:37]: will do. Ideally the github actions should use the makefile so we DRY this up.

File: test/test_notecard.py (hunk 3 above)
Comment (m-mcgowan) [reply test/test_notecard.py:410]: Sure. there's definitely more refactoring we could do, I just wanted to do the minimal to get it working.

File: test/test_notecard.py (hunk 1 above)
Comment (bsatrom) [reply test/test_notecard.py]: What's a way we can do this without adding a second unit testing framework?

**Comment by m-mcgowan** [comment]
I didn't build this specifically to target HIL, I just wanted to factor out the communications channel, which is orthogonal to most of the tests (i.e. the tests are comms-neutral - they don't care about what channel they are using.) It just happens that with that done, adding HIL later will be easier because we can reuse the same mechanism.

We want a simple way to run all the comms-neutral tests over both mocked serial and i2c, and IMHO this was the simplest way to do it without causing maintenance headaches. The only other way I can think of is to duplicate (copy and paste) all the existing serial tests, and rework them to use i2c, but then we have a lot of duplication of code, which will only become worse when we add more comms channels to the test suite.

File: test/test_notecard.py (hunk 1 above)
Comment (m-mcgowan) [reply test/test_notecard.py]: just a fyi - `unittest` was already being used for [mocks](https://github.com/blues/note-python/blame/main/test/test_notecard.py#L7) before this PR. 

`unittest` is built into python so we could rework the pytest , which at present are just assertions for raised exceptions. But I wouldn't stop using pytest - it's a great test runner, while `unittest` deals with expressing tests and creating mocks. 

`pytest` is intended to be used with `unittest` - https://docs.pytest.org/en/7.1.x/how-to/unittest.html#:~:text=pytest%20supports%20running%20Python%20unittest,full%20advantage%20of%20pytest's%20features.

I've reworked the test classes so they don't have `unittest.TestCase` as a base class, and pytest is fine with that, so long as the class name starts with `Test`.

**Comment by bsatrom** [comment]
My apologies @m-mcgowan, the combination of an impenetrable diff and mention of HiL caused me to (wrongly) assume that more was going on here than mocking I2C for testing.

**Review by bsatrom — APPROVED** [review]

**Comment by m-mcgowan** [comment]
No problem! I was going to suggest you view the files side-by-side because then it's much easier to see the diff than what github gives you

**Review by haydenroche5 — APPROVED** [review]

**Pull request merged**